File: app/lib/core/models/promotion.dart
enum PromotionType { banner, interstitial, both }

class Promotion {
  final String id;
  final String imageUrl;
  final String title;
  final String subtitle;
  final String? actionUrl;
  final PromotionType type;
  final DateTime startDate;
  final DateTime endDate;
  final bool isActive;

  const Promotion({
    required this.id,
    required this.imageUrl,
    required this.title,
    required this.subtitle,
    this.actionUrl,
    required this.type,
    required this.startDate,
    required this.endDate,
    required this.isActive,
  });

  factory Promotion.fromJson(Map<String, dynamic> json) => Promotion(
        id: json['id'] as String,
        imageUrl: json['imageUrl'] as String,
        title: json['title'] as String,
        subtitle: json['subtitle'] as String,
        actionUrl: json['actionUrl'] as String?,
        type: PromotionType.values.firstWhere(
          (e) => e.name == (json['type'] as String? ?? 'banner'),
          orElse: () => PromotionType.banner,
        ),
        startDate: DateTime.parse(json['startDate'] as String),
        endDate: DateTime.parse(json['endDate'] as String),
        isActive: json['isActive'] as bool? ?? true,
      );

  bool get isCurrentlyActive {
    final now = DateTime.now();
    return isActive && now.isAfter(startDate) && now.isBefore(endDate);
  }
}

File: backend-express/src/models/Promotion.js
const mongoose = require('mongoose');

const promotionSchema = new mongoose.Schema(
  {
    imageUrl: { type: String, default: '' },
    title: { type: String, required: true },
    subtitle: { type: String, required: true },
    actionUrl: { type: String, default: null },
    type: {
      type: String,
      enum: ['banner', 'interstitial', 'both'],
      default: 'both',
    },
    startDate: { type: Date, required: true },
    endDate: { type: Date, required: true },
    isActive: { type: Boolean, default: true },
    priority: { type: Number, default: 0 }, // higher = shown first
  },
  { timestamps: true }
);

promotionSchema.index({ startDate: 1, endDate: 1, isActive: 1 });

module.exports = mongoose.model('Promotion', promotionSchema);
